geode-core/src/main/java/org/apache/geode/internal/monitoring/ThreadsMonitoringImpl.java

```diff
@@ -96,39 +98,60 @@ public void updateThreadStatus() {
 
   @Override
   public boolean startMonitor(Mode mode) {
-    AbstractExecutor absExtgroup;
+    return startMonitoring(createAbstractExecutor(mode));
+  }
+
+  @Override
+  public void endMonitor() {
+    this.monitorMap.remove(Thread.currentThread().getId());
+  }
+
+  @VisibleForTesting
+  boolean isMonitoring() {
+    return monitorMap.containsKey(Thread.currentThread().getId());
+  }
+
+  @Override
+  public AbstractExecutor createAbstractExecutor(Mode mode) {
     switch (mode) {
       case FunctionExecutor:
-        absExtgroup = new FunctionExecutionPooledExecutorGroup(this);
-        break;
+        return new FunctionExecutionPooledExecutorGroup();
       case PooledExecutor:
-        absExtgroup = new PooledExecutorGroup(this);
-        break;
+        return new PooledExecutorGroup();
       case SerialQueuedExecutor:
-        absExtgroup = new SerialQueuedExecutorGroup(this);
-        break;
+        return new SerialQueuedExecutorGroup();
       case OneTaskOnlyExecutor:
-        absExtgroup = new OneTaskOnlyExecutorGroup(this);
-        break;
+        return new OneTaskOnlyExecutorGroup();
       case ScheduledThreadExecutor:
-        absExtgroup = new ScheduledThreadPoolExecutorWKAGroup(this);
-        break;
+        return new ScheduledThreadPoolExecutorWKAGroup();
       case AGSExecutor:
-        absExtgroup = new GatewaySenderEventProcessorGroup(this);
-        break;
+        return new GatewaySenderEventProcessorGroup();
+      case P2PReaderExecutor:
+        return new P2PReaderExecutorGroup();
       default:
-        return false;
+        throw new IllegalStateException("Unhandled mode=" + mode);
     }
-    this.monitorMap.put(Thread.currentThread().getId(), absExtgroup);
+  }
+
+  @Override
+  public boolean startMonitoring(AbstractExecutor executor) {
+    executor.setStartTime(System.currentTimeMillis());
```

Comment:
I'm a little concerned about the impact of taking yet another millisecond clock reading on every message that we dispatch.  This should be run through extensive performance testing before merging it to develop.

Reply:
I had the same thought. I was even concerned about constantly adding and removing from the concurrent map. We could just use the frequency of the monitor thread to do the time work. For example if it finds an AbstractExecutor in the map that it has not seen before then the monitor thread could set the time it first saw it. If later monitor samples see it again it could use this time (i.e. the one the monitor set) to determine if it is stuck. This could even be done without timestamps. Just a simple counter the monitor incs and the p2p reader thread clears. If the monitor thread is waking up at a fixed interval then you know the thread has been stuck for at least that much time. I like your idea of having custom code in P2PReaderExecutorGroup for this. When I was thinking about it I thought I would need to change the behavior of all the existing threads being monitored. Thanks for the feedback!

Reply:
I converted back to draft mode. Will rework this and submit it again.

Reply:
having the monitor thread set the time is a good idea.  Another thing you could do is have the Monitor thread record the current time when it wakes up and use that for new entries in the map.  GMSHealthMonitor's monitor thread does something like that.

Reply:
the p2p reader now simply sets a volatile boolean before and after messageDispatch.
the startTime for a p2p reader will now be set by the monitor thread.
Let me know what you think